[Autotuner] CI Smoke Test - Tune & Sweep

luarss opened the pull request:
@vvbandeira 
Goal is to do smoke test using CLI function call. 
Supports all platforms: asap7, sky130hd, ihp-sg13g2, gf180

- [x] Tune 
- [x] Sweep
- [x] Change Jenkinsfile and test_helper.sh

luarss committed:
scaffold for smoke tests

Signed-off-by: Jack Luar <39641663+luarss@users.noreply.github.com>

luarss committed:
working smoke test for tune

Signed-off-by: Jack Luar <39641663+luarss@users.noreply.github.com>

luarss committed:
Woring smoke test tune

Signed-off-by: Jack Luar <39641663+luarss@users.noreply.github.com>

rovinski commented on tools/AutoTuner/.gitignore:
How much of this is actually needed?

luarss replied:
It is a generic Python .gitignore, helped to catch some stray files while I was generating the autotuner python package. Will probably remove it.

rovinski commented on tools/AutoTuner/src/autotuner/__init__.py:
empty file?

luarss replied:
I included it because there was a discussion whether to make it a proper python package. Will be removing it for now.

rovinski commented on tools/AutoTuner/README.md:
Why link?

luarss replied:
Better visibility? The old documentation points to the the remote setup of Autotuner.

luarss committed:
Fix smoke tests

Signed-off-by: Jack Luar <39641663+luarss@users.noreply.github.com>

luarss committed:
Remove python packaging code/docs

Signed-off-by: Jack Luar <39641663+luarss@users.noreply.github.com>

luarss committed:
Add CI integration

Signed-off-by: Jack Luar <39641663+luarss@users.noreply.github.com>

luarss committed:
remove blocked platforms for security scan

Signed-off-by: Jack Luar <39641663+luarss@users.noreply.github.com>

luarss committed:
Add dependency install to test (will remove later once PR merged)

Signed-off-by: Jack Luar <39641663+luarss@users.noreply.github.com>

luarss committed:
fix platform upper case

Signed-off-by: Jack Luar <39641663+luarss@users.noreply.github.com>

vvbandeira commented on tools/AutoTuner/src/autotuner/distributed.py:
```diff
@@ -960,5 +961,8 @@ def sweep():
             sys.exit(1)
+
+        # if successful
+        sys.exit(0)
```
Is this required?

luarss replied:
```suggestion
```

vvbandeira commented on tools/AutoTuner/test/smoke_test_sweep.py:
```diff
@@ -0,0 +1,74 @@
+#     def test_sweep(self):
+#         out = subprocess.run(self.command, shell=True, check=True)
+#         successful = out.returncode == 0
+#         self.assertTrue(successful)
```
Remove code instead of commenting it out.

vvbandeira commented on tools/AutoTuner/test/smoke_test_tune.py:
```diff
@@ -0,0 +1,60 @@
+#     def test_tune(self):
+#         out = subprocess.run(self.command, shell=True, check=True)
+#         successful = out.returncode == 0
+#         self.assertTrue(successful)
```
Remove code instead of commenting it out.

vvbandeira commented on docs/user/InstructionsForAutoTuner.md:
````diff
@@ -193,6 +193,14 @@ We show three different views possible at the end, namely: `Table View`, `Scatte
+cd tools/AutoTuner/tests
+python3 smoke_test_tune.py
+python3 smoke_test_sweep.py
+```
````
Shouldn't there be instructions on the virtual env setup?

luarss committed:
address comments

Signed-off-by: Jack Luar <39641663+luarss@users.noreply.github.com>

luarss committed:
fix gcd unstable and genreport

Signed-off-by: luarss <39641663+luarss@users.noreply.github.com>

luarss committed:
hotfix for autotuner report files

Signed-off-by: luarss <39641663+luarss@users.noreply.github.com>

luarss committed:
add ihp gcd params

Signed-off-by: Jack Luar <39641663+luarss@users.noreply.github.com>

vvbandeira commented on tools/AutoTuner/src/autotuner/distributed.py:
```diff
@@ -882,6 +882,7 @@ def sweep():
     print('[INFO TUN-0009] Waiting for results.')
     ray.get(workers)
     print('[INFO TUN-0010] Sweep complete.')
+    sys.exit(0)
```
This should not be needed as the script will end after the call of `sweep()`. We should only exit with an error in places other than `__main__.`

vvbandeira commented on docs/user/InstructionsForAutoTuner.md:
```diff
@@ -30,7 +30,7 @@ which is a free minimal installer for the package manager `conda`.
+conda create -n autotuner_env python=3.8 -y
 conda activate autotuner_env
 
 # install requirements
```
A simple `setup.sh` would be helpful here. Users can use a single command to set up, and CI can use the same.

luarss replied:
`. autotuner_env/bin/activate`

vvbandeira commented on docs/user/InstructionsForAutoTuner.md:
```diff
@@ -193,6 +193,17 @@ We show three different views possible at the end, namely: `Table View`, `Scatte
+# Smoke test
+cd tools/AutoTuner/tests
+python3 smoke_test_tune.py
+python3 smoke_test_sweep.py
```
Could the tests themselves call the `setup.sh` (as suggested above) to activate the environment?

luarss replied:
They can, but from the same principles of ORFS can we leave it optional? Similar to how we have `Env.sh` as optional

luarss replied:
Update the docs here. 3 step instructions
- install depen
- source env
- run test.

luarss committed:
fix comments

Signed-off-by: luarss <39641663+luarss@users.noreply.github.com>

luarss committed:
fix doc clarity comments

Signed-off-by: luarss <39641663+luarss@users.noreply.github.com>

vvbandeira commented on flow/test/test_helper.sh:
```diff
@@ -63,4 +63,36 @@ if [ ! -z ${MAKE_ISSUE+x} ]; then
+fi
+
+if [ $RUN_AUTOTUNER -eq 1 ]; then
+  # change directory to ../
```
```suggestion
  # change directory to the root of the repo
```

vvbandeira committed:
Update flow/test/test_helper.sh

Signed-off-by: Vitor Bandeira <vvbandeira@users.noreply.github.com>

vvbandeira commented on flow/test/test_helper.sh:
```diff
@@ -63,4 +63,36 @@ if [ ! -z ${MAKE_ISSUE+x} ]; then
+  
+  echo "Running Autotuner smoke sweep test"
+  python3 -m unittest tools.AutoTuner.test.smoke_test_sweep.${PLATFORM^^}SweepSmokeTest.test_sweep
+fi
```
``` suggestion
if [ $RUN_AUTOTUNER -eq 1 ]; then
  # change directory to the root of the repo
  cd ..
  echo "Install dependencies in Venv"
  ./tools/AutoTuner/installer.sh
  ./tools/AutoTuner/setup.sh

  # remove dashes
  PLATFORM=${PLATFORM//-/}
  # convert to uppercase
  PLATFORM=${PLATFORM^^}

  echo "Running Autotuner smoke tune test"
  python3 -m unittest tools.AutoTuner.test.smoke_test_tune.${PLATFORM}TuneSmokeTest.test_tune

  echo "Running Autotuner smoke sweep test"
  python3 -m unittest tools.AutoTuner.test.smoke_test_sweep.${PLATFORM}SweepSmokeTest.test_sweep
fi
```

vvbandeira committed:
Update flow/test/test_helper.sh

Signed-off-by: Vitor Bandeira <vvbandeira@users.noreply.github.com>

vvbandeira commented on tools/AutoTuner/setup.sh:
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
+# Get the directory of the current script
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
```
```suggestion
SCRIPT_DIR="$(dirname "${BASH_SOURCE[0]}")"
```

vvbandeira committed:
Update tools/AutoTuner/setup.sh

Signed-off-by: Vitor Bandeira <vvbandeira@users.noreply.github.com>

luarss committed:
fix black

Signed-off-by: luarss <39641663+luarss@users.noreply.github.com>

luarss committed:
no need cd ../

Signed-off-by: luarss <39641663+luarss@users.noreply.github.com>

luarss committed:
Fix chmod x issues

Signed-off-by: Jack Luar <39641663+luarss@users.noreply.github.com>

luarss committed:
add debug print

Signed-off-by: Jack Luar <39641663+luarss@users.noreply.github.com>

luarss committed:
test copy tools/AutoTuner

Signed-off-by: Jack Luar <39641663+luarss@users.noreply.github.com>

vvbandeira commented on tools/AutoTuner/.gitignore:
Please remove comments and clean up the list to only include what is required for our use case.

vvbandeira commented on tools/AutoTuner/test/smoke_test_sweep.py:
```diff
@@ -0,0 +1,75 @@
+            f" --experiment {self.experiment}"
+            f" --config {self.config}"
+            f" --jobs {self.jobs}"
+            f" sweep"
+        )
+
```
Add the `test_sweep` method and throw an exception in the body saying that this needs to be implemented in the derivative classes.

vvbandeira commented on tools/AutoTuner/test/smoke_test_tune.py:
```diff
@@ -0,0 +1,61 @@
+            f" --platform {self.platform}"
+            f" --experiment {self.experiment}"
+            f" --config {self.config}"
+            f" tune --samples 1"
+        )
+
```
Same as above.

vvbandeira commented on tools/AutoTuner/.gitignore:
```diff
@@ -1,2 +1,164 @@
+#  option (not recommended) you can uncomment the following to ignore the entire idea folder.
+#.idea/
+
+>>>>>>> e0148686 (working smoke test for tune)
```
Remove?

luarss committed:
slimmer gitignore and add test_xx functions for the base class

Signed-off-by: luarss <39641663+luarss@users.noreply.github.com>

vvbandeira commented on docker/Dockerfile.builder:
```diff
@@ -33,6 +33,7 @@ COPY --link flow/platforms flow/platforms
 COPY --link flow/util flow/util
 COPY --link flow/scripts flow/scripts
 COPY --link flow/designs flow/designs
+COPY -- link tools/AutoTuner tools/AutoTuner
```
```suggestion
COPY --link tools/AutoTuner tools/AutoTuner
```

vvbandeira reviewed (changes requested):
Looks like you had a typo

vvbandeira committed:
Update docker/Dockerfile.builder

Signed-off-by: Vitor Bandeira <vvbandeira@users.noreply.github.com>

vvbandeira commented on tools/AutoTuner/installer.sh:
Why do we need this empty file?

luarss replied:
Not an empty file, I was just making it executable (chmod +x)

luarss committed:
fix py syntax and dockerfile

Signed-off-by: luarss <jluar@precisioninno.com>

vvbandeira committed:
ci: bump version

The new version makes sure to copy the required info into the container
so that AT CI can run correctly.

Signed-off-by: Vitor Bandeira <vvbandeira@precisioninno.com>

vvbandeira commented on docker/Dockerfile.builder:
```diff
@@ -33,10 +33,12 @@ COPY --link flow/platforms flow/platforms
 COPY --link \
-     --exclude=.git* --exclude=tools/ --exclude=docs/ --exclude=docker/ \
+     --exclude=.git* --exclude=tools/OpenROAD --exclude=tools/codespace \
+     --exclude=tools/yosys --exclude=tools/yosys_utils --exclude=docs/ --exclude=docker/ \
```
I think we can revert this change, AT is already copied in line 36.

luarss committed:
revert Dockerfile and fix installer

Signed-off-by: luarss <jluar@precisioninno.com>

vvbandeira commented on tools/AutoTuner/installer.sh:
```diff
@@ -6,6 +6,7 @@ cd $script_dir
 
 # Define the virtual environment name
 venv_name="autotuner_env"
+required_version="3.8"
```
Didn't we agree somewhere to drop the specific version and keep with either `python` or `python3`?

luarss replied:
`python` did not resolve to any command in previous iteration, maybe I will do 

```
required_version="3"
python${required_version} ...
```

luarss committed:
revert Dockerfile and change installer to use py3

Signed-off-by: luarss <jluar@precisioninno.com>

luarss committed:
No more cd during installer, standardise setup script and add gitignore

Signed-off-by: luarss <jluar@precisioninno.com>

luarss committed:
source setup script

Signed-off-by: luarss <jluar@precisioninno.com>

luarss committed:
add PATH to export

Signed-off-by: luarss <jluar@precisioninno.com>

vvbandeira commented on tools/AutoTuner/installer.sh:
```diff
@@ -2,11 +2,11 @@
 # Define the virtual environment name
 venv_name="autotuner_env"
+required_version="3"
 python${required_version} -m venv "$script_dir/$venv_name"
```
```suggestion
python3 -m venv "$script_dir/$venv_name"
```

vvbandeira commented on flow/test/test_helper.sh:
```diff
@@ -77,4 +77,38 @@ if [ $ret -eq 0 ] && grep -q 'power:' <(echo $TARGETS); then
+        RUN_AUTOTUNER=0
+        ;;
+  esac
+fi
```
```suggestion
# Run Autotuner CI specifically for gcd on selected platforms.
RUN_AUTOTUNER=0
case $DESIGN_NAME in
  "gcd")
    RUN_AUTOTUNER=1
    ;;
esac
case $PLATFORM in
     "asap7" | "sky130hd" | "ihp-sg13g2" )
      # Keep RUN_AUTOTUNER enabled only for these platforms
      ;;
     *)
      RUN_AUTOTUNER=0
      ;;
esac
```

vvbandeira committed:
Apply suggestions from code review

Signed-off-by: Vitor Bandeira <vvbandeira@users.noreply.github.com>
